framework/src/outputs/Checkpoint.C

```diff
@@ -75,60 +77,74 @@ Checkpoint::directory() const
 
-  // Write the checkpoint file
-  io.write(curr_file_struct.checkpoint);
+    curr_file_struct.checkpoint = current_file + getMeshFileSuffix(_binary);
+    curr_file_struct.system = current_file + _restartable_data_io.getESFileExtension(_binary);
+    curr_file_struct.restart = current_file + _restartable_data_io.getRestartableDataExt();
 
-  // Write out the restartable mesh meta data if there is any (only on processor zero)
-  if (processor_id() == 0)
-  {
-    for (auto & map_pair :
-         libMesh::as_range(_app.getRestartableDataMapBegin(), _app.getRestartableDataMapEnd()))
-    {
-      const RestartableDataMap & meta_data = map_pair.second.first;
-      const std::string & suffix = map_pair.second.second;
-      const std::string filename(curr_file_struct.checkpoint + "/meta_data" + suffix +
-                                 _restartable_data_io.getRestartableDataExt());
+    // Write the checkpoint file
+    io.write(curr_file_struct.checkpoint);
 
-      curr_file_struct.restart_meta_data.emplace(filename);
-      _restartable_data_io.writeRestartableData(filename, meta_data);
+    // Write out the restartable mesh meta data if there is any (only on processor zero)
+    if (processor_id() == 0)
+    {
+      for (auto & map_pair :
+           libMesh::as_range(_app.getRestartableDataMapBegin(), _app.getRestartableDataMapEnd()))
+      {
+        const RestartableDataMap & meta_data = map_pair.second.first;
+        const std::string & suffix = map_pair.second.second;
+        const std::string filename(curr_file_struct.checkpoint + "/meta_data" + suffix +
+                                   _restartable_data_io.getRestartableDataExt());
+
+        curr_file_struct.restart_meta_data.emplace(filename);
+        _restartable_data_io.writeRestartableData(filename, meta_data);
+      }
     }
-  }
 
-  // Write the system data, using ENCODE vs WRITE based on ascii vs binary format
-  _es_ptr->write(curr_file_struct.system,
-                 EquationSystems::WRITE_DATA | EquationSystems::WRITE_ADDITIONAL_DATA |
-                     EquationSystems::WRITE_PARALLEL_FILES,
-                 renumber);
+    // Write the system data, using ENCODE vs WRITE based on ascii vs binary format
+    _es_ptr->write(curr_file_struct.system,
+                   EquationSystems::WRITE_DATA | EquationSystems::WRITE_ADDITIONAL_DATA |
+                       EquationSystems::WRITE_PARALLEL_FILES,
+                   renumber);
+
+    // Write the restartable data
+    _restartable_data_io.writeRestartableDataPerProc(curr_file_struct.restart, _restartable_data);
 
-  // Write the restartable data
-  _restartable_data_io.writeRestartableDataPerProc(curr_file_struct.restart, _restartable_data);
+    // Remove old checkpoint files
+    updateCheckpointFiles(curr_file_struct);
+
+    // Stop outputting the checkpoint if this is a signalled checkpoint
+    if (Moose::autosave_flag){
+      _should_output = false;
+      Moose::autosave_flag = false;
```

Comment:
Do you need two flags? You could make the argument if you really think users need to control this in addition to the signaling flag. It's not clear that's your intent though.

Reply:
I think so. I don't think its possible to control the flow between a normal checkpoint object and the signal-based one without the double flags. `_should_output` is set to false in autoCheckpointAction to prevent it from writing the signal checkpoint until a signal is received, which is based on the global `autosave_flag` var.